tests/abi/tools/revng-check-compatibility-with-abi/Verify.cpp

```diff
@@ -230,114 +227,299 @@ void VH::verify(const abi::FunctionType::Layout &FunctionLayout,
                             "expected.";
       revng_abort(Error.c_str());
     }
+  }
+}
+using VH = VerificationHelper;
+void VH::returnValue(const abi::FunctionType::Layout &FunctionLayout,
+                     const RVTest &Iterations) const {
+  auto PointerSize = model::Architecture::getPointerSize(Architecture);
+
+  for (const abi::runtime_test::ReturnValueTest &Test : *Iterations) {
+    const auto &Found = Test.ReturnValue;
+    const auto &Expected = Test.ExpectedReturnValue;
+    if (Found.Bytes.size() != Expected.Bytes.size()) {
+      std::string Error = "Return value size of '" + FunctionName.str()
+                          + "' is not consistent: something is *very* wrong "
+                            "with the test toolchain.";
+      revng_abort(Error.c_str());
+    }
 
-    // Check the return value.
-    auto RVR = FunctionLayout.returnValueRegisters();
-    auto ReturnValueRegisterList = ABI.sortReturnValues(std::move(RVR));
-    if (!Helper.Iteration.ReturnValue.Bytes.empty()) {
-      if (ReturnValueRegisterList.size() == 0) {
-        std::string Error = "Verification of the return value of '"
+    std::uint64_t MatchingByteCount = 0;
+    for (auto [FByte, EByte] : llvm::zip(Found.Bytes, Expected.Bytes))
+      if (FByte == EByte)
+        ++MatchingByteCount;
+
+    // Sadly, we cannot do a full comparison since return value structs often
+    // contain padding, value of which is not guaranteed to be preserved,
+    // causing the test to fail. We either need a way to tell if a specific byte
+    // belongs to the padding or not (which would restrict stuff we can test,
+    // for example, nested structs) or assume that non-perfect matches are
+    // acceptable.
+    //
+    // We know for sure that maximum padding amount is
+    // "maximum_possible_alignment_for_the_architecture - 1", which is achieved
+    // when a struct like this is returned:
+    // ```cpp
+    // struct ReturnValue {
+    //   field_with_maximum_alignment a; // let's say the alignment is 4.
+    //   field_with_minimum_alignment b; // the alignment always is 1.
+    //   // invisible 3 bytes here (4 - 1).
+    // };
+    // ```
+    // Which adds up to 3/8 (~37.5%). In the case where maximum alignment is
+    // 16 bytes - this jumps to 15/32 (~46.9%), and so on. Knowing that,
+    // we can assume that, at the very least, _more_ than half of the bytes
+    // will match.
+    //
+    // Also, considering we do multiple iterations of each test, the chance
+    // of having false positives is extremely low as well.
```

Comment:
Reading this I cannot understand much.
Make an explicit, specific example referencing a specific architecture, ABI, return value.

Reply:
How does something like this sound?
```cpp
  // Sadly, we cannot do a full comparison since return value structs often
  // contain padding, value of which is not guaranteed to be preserved,
  // causing the test to fail. We either need a way to tell if a specific byte
  // belongs to the padding or not (which would restrict stuff we can test,
  // for example, nested structs) or assume that non-perfect matches are
  // acceptable.
  // An example of a struct that exhibits this behaviour (i386):
  // struct ReturnValue {
  //   uint8_t b; // _Alignof(uint8_t) == 1
  //   // 3 invisible bytes here
  //   // (_Alignof(ReturnValue) - _Alignof(uint8_t) = 4 - 1 = 3)
  //   uint32_t a; // _Alignof(uint32_t) == 4
  // };
  //
  // So, when looking at the return value we detect, it looks something like
  // `[ 0x11, 0xXX, 0xXX, 0xXX, 0x22, 0x33, 0x44, 0x55 ]`
  // where `0xXX` is unuinitialized value that is not stable and changes from
  // execution to execution.
  //
  // But, luckily, by the definition of how such padding is introduced, we are
  // guaranteed that in the worst case it's going to occupy `half_the_size - 1`
  // bytes. So, we can at least enforce that. And, thanks to doing multiple
  // iterations for each tests, the chance of false positives is extremely low.
  ```